components/AnimatedImage/demo/src/jvmMain/kotlin/org/jetbrains/compose/animatedimage/demo/Main.kt

```diff
@@ -0,0 +1,60 @@
+package org.jetbrains.compose.animatedimage.demo
+
+import androidx.compose.foundation.Image
+import androidx.compose.foundation.layout.Column
+import androidx.compose.material.Button
+import androidx.compose.material.CircularProgressIndicator
+import androidx.compose.material.Text
+import androidx.compose.runtime.*
+import androidx.compose.ui.window.singleWindowApplication
+import org.jetbrains.compose.animatedimage.*
+
+fun main() = singleWindowApplication {
+    val url =
+        "https://raw.githubusercontent.com/JetBrains/skija/ccf303ebcf926e5ef000fc42d1a6b5b7f1e0b2b5/examples/scenes/images/codecs/animated.gif"
+
+    // Preloaded image that will be shown when the user clicks on the button
+    val preloadedImage = remember { mutableStateOf<AnimatedImage?>(null) }
+
+    var showPreloaded by remember { mutableStateOf(false) }
+    val preloadedImageValue = preloadedImage.value
+
+    LaunchedEffect(Unit) {
+        // Load an image sync
+        preloadedImage.value = loadAnimatedImage(url) // use "loadResourceAnimatedImage" for resources
+    }
+
+    // Load an image async
+    val asyncAnimatedImage =
+        asyncAnimatedImageLoaderState(url)  // use "asyncResourceAnimatedImageLoaderState" for resources
```

Comment:
I am a little worried about this name being too long. What if we create a common function for loading something?

```
    // Load an image async
    val image = load { loadAnimatedImage(url) }  // use "load { loadResourceAnimatedImage(url) }" for resources

    Column {
        Image(
            loadOrNull { loadAnimatedImage(url) }?.animate() ?: ImageBitmap.Blank,
            contentDescription = null,
            Modifier.size(100.dp)
        )

        when (image) {
            is LoadState.Success -> Image(
                bitmap = image.value.animate(),
                contentDescription = null,
            )
            is LoadState.Loading -> CircularProgressIndicator()
            is LoadState.Error -> Text("Error!")
        }

        ...
    }
```

Also, we can rename this library to `resources`, and put common loading logic for Android/desktop/web/etc in the future (and when it will be ready for core Compose, we can move it there, and deprecate here):
```
package org.jetbrains.compose.resources

sealed class LoadState<T> {
    class Loading<T> : LoadState<T>()
    data class Success<T>(val value: T) : LoadState<T>()
    data class Error<T>(val exception: Exception) : LoadState<T>()
}

@Composable
fun <T> load(load: suspend () -> T): LoadState<T> {
    var state: LoadState<T> by remember(load) { mutableStateOf(LoadState.Loading()) }
    LaunchedEffect(load) {
        state = try {
            LoadState.Success(load())
        } catch (e: Exception) {
            LoadState.Error(e)
        }
    }
    return state
}

@Composable
fun <T: Any> loadOrNull(load: suspend () -> T): T? {
    val state = load(load)
    return (state as? LoadState.Success<T>)?.value
}

private val BlankBitmap = ImageBitmap(1, 1)
val ImageBitmap.Companion.Blank get() = BlankBitmap

@Composable
fun AnimatedImage.animate(): ImageBitmap {
    ...
}

// AnimatedImageLoader can be private in this case
```

We plan to do it anyway, so if you help now, we appreciate it :)

Reply:
Agreed with the naming being too long, that's why I wasn't sure about it.

Making it as a generic method would be great to make it available to everyone + fix the naming issue, lgtm. I'll update the PR with the suggestions.

However, I am not sure about renaming the lib to `resources`.  It's great for the loading methods but the animated image methods seem a bit off here, as it's something used in a very specific feature that most people don't expect to find in a lib with such name.

What do you think about having 2 different libs, one for the resources and the other for the animated image stuff?

Reply:
> However, I am not sure about renaming the lib to resources

Okay, let's create two separate libraries for now. If we change decision, we can deprecate `animatedimage` library and move it to `resources` library. By the way, could you add 'animatedImage` to [this](https://github.com/JetBrains/compose-jb/blob/f281b75f7731269065cdaa861bed73ef2c24373e/gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/ComposePlugin.kt#L198) place (`resources` let's keep "private" for now, only as "api" dependency of "animatedImage")?

The plan is to put in `resources` things, which help to load some abstract resources from external/internal sources in commonMain code. Maybe to put some tooling into this library (for example, pack internal resources as part of native distribution). It is also possible, that we will add ability to load concrete resources (images/animations/fonts/strings). But it is very early concept, so I agree, let's don't confuse people, putting only support for animations.

Reply:
I've pushed the changes, let me know what do you think.